cxx-qt-lib/src/qdatetime.rs

```diff
@@ -0,0 +1,267 @@
+        // function call safe.
+        let date = unsafe { qdatetime_date(self) };
+
+        // Safety:
+        //
+        // Given that a QDateTime can only be constructed using [`let_qdatetime!`] macro,
+        // it is safe to assume that self is a valid QDateTime reference which makes this
+        // function call safe.
+        let time = unsafe { qdatetime_time(self) };
+
+        DateTime::from_date_time(date, time)
+    }
+}
+
+#[doc(hidden)]
+#[repr(C)]
+pub struct StackQDateTime {
+    // Static assertions in cxx_qt.cpp validate that this
+    // is large enough and aligned enough.
+    space: MaybeUninit<[usize; 1]>,
+}
+
+#[allow(missing_docs)]
+impl StackQDateTime {
+    pub fn new() -> Self {
+        StackQDateTime {
+            space: MaybeUninit::uninit(),
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Calling this function twice on the same StackQDateTime is unsafe
+    /// and leads to undefined behaviour. It is therefore recommended
+    /// to not use this function directly and instead use the [`let_qdatetime!`]
+    /// macro which ensures that safe behaviour.
+    pub unsafe fn init(&mut self, datetime: &DateTime) -> Pin<&mut QDateTime> {
+        let this = &mut *self.space.as_mut_ptr().cast::<MaybeUninit<QDateTime>>();
+        let value = datetime.deref();
+        qdatetime_init(this, &value.date, &value.time);
+        Pin::new_unchecked(&mut *this.as_mut_ptr())
+    }
+}
+
+impl Default for StackQDateTime {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Drop for StackQDateTime {
+    fn drop(&mut self) {
+        // # Safety
+        //
+        // This simply calls ~QDateTime on self.space which is safe as long
+        // as self.space contains a valid QDateTime. Using the [`let_qdatetime!`]
+        // macro guarantees that this will be the case.
+        unsafe {
+            let this = &mut *self.space.as_mut_ptr().cast::<MaybeUninit<QDateTime>>();
+            qdatetime_drop(this);
+        }
+    }
+}
+
+// Safety:
+//
+// The code in this file ensures that QDateTime can only ever be allocated
+// on the stack in pinned form which avoids the pitfalls of trying to
+// move this type that has a non-trivial move constructor.
+unsafe impl ExternType for QDateTime {
+    type Id = type_id!("QDateTime");
+    type Kind = cxx::kind::Opaque;
+}
+
+pub struct DateTimeImpl {
+    date: QDate,
+    time: QTime,
+}
+
+impl DateTimeImpl {
+    pub fn new(date: QDate, time: QTime) -> Self {
+        Self { date, time }
+    }
+}
+
+#[repr(C)]
+pub struct DateTime {
+    pub d: Box<DateTimeImpl>,
+}
+
+impl DateTime {
+    pub fn from_date_time(date: QDate, time: QTime) -> Self {
+        Self {
+            d: Box::new(DateTimeImpl { date, time }),
+        }
+    }
+
+    pub fn date(&self) -> &QDate {
+        &self.d.date
+    }
+
+    pub fn time(&self) -> &QTime {
+        &self.d.time
+    }
+
+    pub fn set_date(&mut self, date: QDate) {
```

Comment:
Should we allow mutating of this DateTime or should they always need to construct a new one ?